File: dart/bucket-sort.dart
List<int> bucketSort(List<int> arr, [int bucketSize = 5]) {
  if (arr.isEmpty) return arr;

  int min = arr.reduce((curr, next) => curr < next ? curr : next);
  int max = arr.reduce((curr, next) => curr > next ? curr : next);

  int bucketCount = ((max - min) / bucketSize).floor() + 1;
  List<List<int>> buckets = List.generate(bucketCount, (i) => []);

  for (int i = 0; i < arr.length; i++) {
    int bucketIndex = ((arr[i] - min) / bucketSize).floor();
    buckets[bucketIndex].add(arr[i]);
  }

  int index = 0;
  for (List<int> bucket in buckets) {
    bucket.sort();
    for (int num in bucket) {
      arr[index++] = num;
    }
  }

  return arr;
}

void main() {
  List<int> arr = [5, 2, 9, 1, 5, 6];
  print(bucketSort(arr));
}

File: javascript/bucket-sort.js
function bucketSort(arr, bucketSize = 5) {
    if (arr.length === 0) return arr;

    let min = Math.min(...arr);
    let max = Math.max(...arr);

    let bucketCount = Math.floor((max - min) / bucketSize) + 1;
    let buckets = Array.from({ length: bucketCount }, () => []);

    for (let i = 0; i < arr.length; i++) {
        let bucketIndex = Math.floor((arr[i] - min) / bucketSize);
        buckets[bucketIndex].push(arr[i]);
    }

    return buckets.reduce((sortedArr, bucket) => {
        return sortedArr.concat(bucket.sort((a, b) => a - b));
    }, []);
}

let arr = [5, 2, 9, 1, 5, 6];
console.log(bucketSort(arr));
